### PR TITLE
Feature/from netty to apachehttpclient

### PR DESCRIPTION
**Pull request by mat1e**
Replace #899

I purpose to change transport layer from Netty to ApacheHttpClient.

It has been developed to fix [jenkins-docker-timeout-issue](https://github.com/groupe-edf/jenkins-docker-timeout-issue) that come with netty implmentation in docker-java.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Review by pjdarton — CHANGES_REQUESTED** [review]

File: pom.xml
Comment (mat1e) [reply pom.xml:77]: I agree. Updates of docker-java-api-plugin are requested in [jenkinsci/docker-java-api-plugin/pull/26](https://github.com/jenkinsci/docker-java-api-plugin/pull/26). This will be removed when the PR will be merged.

File: pom.xml
Comment (MarkEWaite) [inline pom.xml:193]: Please don't add guava as a dependency, especially an older guava version like this.

If you absolutely need the capabilities provided by guava, please review https://www.jenkins.io/blog/2021/11/09/guava-upgrade/ and find a way to rely on the guava that is provided in Jenkins

**Review by MarkEWaite — CHANGES_REQUESTED** [review]

File: pom.xml
Comment (mat1e) [reply pom.xml:193]: Thank you for your answer. 
When I opened this PR, Guava wasn't updated into Jenkins and it was an older version than 19.0. 
I just took the first version that worked with my changes (because 11.0.1 provided before was too old).
I no longer need this dependency now, I can use the new version provided by Jenkins.

File: src/test/java/io/jenkins/docker/connector/DockerComputerConnectorTest.java
Comment (Peter-Darton-i2) [inline src/test/java/io/jenkins/docker/connector/DockerComputerConnectorTest.java:174]: Note that this code change ought to disappear if this PR branch was rebased on main, as the change is already in main.

Comment (mat1e) [reply src/test/java/io/jenkins/docker/connector/DockerComputerConnectorTest.java:174]: I'm checking to do the rebase.

**Comment by Peter-Darton-i2** [comment]
FYI I re-triggered the build (just in case the failure was a one-off) but that doesn't seem to be the case. It looks like the all of the integration tests are reporting:
```
java.lang.NoClassDefFoundError: com/google/common/collect/MultimapBuilder
	at com.github.dockerjava.core.DefaultDockerCmdExecFactory$DefaultWebTarget.<init>(DefaultDockerCmdExecFactory.java:57)
	at com.github.dockerjava.core.DefaultDockerCmdExecFactory.getBaseResource(DefaultDockerCmdExecFactory.java:40)
	at com.github.dockerjava.core.AbstractDockerCmdExecFactory.createListContainersCmdExec(AbstractDockerCmdExecFactory.java:280)
	at com.github.dockerjava.core.DockerClientImpl.listContainersCmd(DockerClientImpl.java:379)
	at io.jenkins.dockerjavaapi.client.DelegatingDockerClient.listContainersCmd(DelegatingDockerClient.java:264)
	at com.nirima.jenkins.plugins.docker.DockerCloud.countContainersInDocker(DockerCloud.java:593)
	at io.jenkins.docker.connector.DockerComputerConnectorTest.dockerIsStillBusy(DockerComputerConnectorTest.java:151)
	at io.jenkins.docker.connector.DockerComputerConnectorTest.cleanup(DockerComputerConnectorTest.java:116)
```

i.e. it looks like dockerjava requires `com.google.common.collect.MultimapBuilder` but it isn't on the classpath.

**Comment by pjdarton** [comment]
FYI I re-triggered the build (just in case the failure was a one-off) but that doesn't seem to be the case. It looks like the all of the integration tests are reporting:
```
java.lang.NoClassDefFoundError: com/google/common/collect/MultimapBuilder
	at com.github.dockerjava.core.DefaultDockerCmdExecFactory$DefaultWebTarget.<init>(DefaultDockerCmdExecFactory.java:57)
	at com.github.dockerjava.core.DefaultDockerCmdExecFactory.getBaseResource(DefaultDockerCmdExecFactory.java:40)
	at com.github.dockerjava.core.AbstractDockerCmdExecFactory.createListContainersCmdExec(AbstractDockerCmdExecFactory.java:280)
	at com.github.dockerjava.core.DockerClientImpl.listContainersCmd(DockerClientImpl.java:379)
	at io.jenkins.dockerjavaapi.client.DelegatingDockerClient.listContainersCmd(DelegatingDockerClient.java:264)
	at com.nirima.jenkins.plugins.docker.DockerCloud.countContainersInDocker(DockerCloud.java:593)
	at io.jenkins.docker.connector.DockerComputerConnectorTest.dockerIsStillBusy(DockerComputerConnectorTest.java:151)
	at io.jenkins.docker.connector.DockerComputerConnectorTest.cleanup(DockerComputerConnectorTest.java:116)
```
i.e. it looks like dockerjava requires `com.google.common.collect.MultimapBuilder` but it isn't on the classpath.

File: pom.xml
Comment (microscotch) [reply pom.xml:83]: Hi @mat1e,

PR jenkinsci/docker-java-api-plugin/pull/26 has been merged on Oct 12, 2022, so will you update this PR to remove this dependency ?

**Comment by microscotch** [comment]
Hi @mat1e,

I think I found a way to fix latest issue with your PR.

For testing purpose, I'm running latest Jenkins LTS (currently 2.375.1) from its ARM64 docker image.
Dockerhost's local docker socket was binded to the Jenkins container in order to setup a docker cloud which use the binded local docker socket.
Unfortunately when I run cloud configuration test I received exception `java.lang.UnsatisfiedLinkError: no netty_transport_native_epoll_aarch_64 in java.library.path.

After googling raised exception, I found [this post](https://github.com/jenkinsci/docker-plugin/issues/866) which redirects me to your PR.

I read the thread and try to fix latest issue about exception `java.lang.NoClassDefFoundError: com/google/common/collect/MultimapBuilder` mentionned by @MarkEWaite in his [last post](https://github.com/jenkinsci/docker-plugin/pull/900#issuecomment-1327735530).

Indeed, this class is part of guava, the dependency he asks you to remove because of [upgrade of guava](https://www.jenkins.io/blog/2021/11/09/guava-upgrade/), therefor to be able to resolve class `com/google/common/collect/MultimapBuilder`, plugin have now to depends at least on Jenkins 2.320 (which bundles guava 31.0.1).

So, I updated pom.xml in order to:
* set maven property `jenkins.version` to 2.230
* remove dependency docker-java-transport-httpclient5:3.2.13 and replace it with docker-java-api:3.2.13-64.v2d092835754e (as requested by @pjdarton [here](https://github.com/jenkinsci/docker-plugin/pull/900#discussion_r894645891)), were your PR https://github.com/jenkinsci/docker-java-api-plugin/pull/26 was merged (but unfortunately not yet released).

Then I've installed [docker-java-api-3.2.13-64.v2d092835754e.hpi](https://ci.jenkins.io/job/Plugins/job/docker-java-api-plugin/job/master/lastSuccessfulBuild/artifact/org/jenkins-ci/plugins/docker-java-api/3.2.13-64.v2d092835754e/docker-java-api-3.2.13-64.v2d092835754e.hpi) and my build result hpi from my Jenkins plugin manager.

After restarting my Jenkins, docker cloud configuration works like a charm, and Jenkins is able to provide docker template as expected :)

Best regards,

Microscotch.

**Comment by microscotch** [comment]
Hi @mat1e,

I've just perform some tests to check my hypothesis, which validate my hypothesis regarding the minimal jenkins version required.

My test protocol was to build plugin with maven test activated:

```
--------------------------------------------------------------------------------------
| Attempt | Jenkins | docker-java-plugin | docker-java-transport-httpclient5 | Tests |
| ------- | ------- | ----------------------- | ---------------------------- | ----- |
| #1      | 2.320   | 3.2.13-64.v2d092835754e | Removed                      | OK    |
| #2      | 2.303.3 | 3.2.13-37.vf3411c9828b9 | 3.2.13                       | KO    |
| #3      | 2.303.3 | 3.2.13-64.v2d092835754e | Removed                      | KO    |
| #4      | 2.320   | 3.2.13-37.vf3411c9828b9 | 3.2.13                       | OK    |
--------------------------------------------------------------------------------------
```

 * `#1` :heavy_check_mark: Suggested fix 
 * `#2` :x: Your PR 
 * `#3` :x: Suggested fix without original minimal Jenkins version required
 * `#4` :heavy_check_mark: Your PR with suggested minimal Jenkins version required upgraded

So your PR can be accepted only by updating maven property `jenkins.version` to at least 2.320.

And once next version of docker-java-plugin will be released, dependency `docker-java-transport-httpclient5` might be alse removed as you mentioned into [this comment](https://github.com/jenkinsci/docker-plugin/pull/900#discussion_r895860635)

**Comment by mat1e** [comment]
Hi @microscotch,
Many thanks for your help! I blocked on the guava ClassNotFoundException...

**Comment by mat1e** [comment]
@pjdarton @MarkEWaite, I need to change the target jenkins version for the build in the JenkinsFile but I don't have the permission to change it as described [here](https://docs.cloudbees.com/docs/cloudbees-ci-kb/latest/client-and-managed-masters/why-jenkinsfile-changes-are-not-reflected-in-pr-build)

**Comment by MarkEWaite** [comment]
> @pjdarton @MarkEWaite, I need to change the target jenkins version for the build in the JenkinsFile but I don't have the permission to change it as described [here](https://docs.cloudbees.com/docs/cloudbees-ci-kb/latest/client-and-managed-masters/why-jenkinsfile-changes-are-not-reflected-in-pr-build)

I launched a "Replay" of the most recent build using the modified Jenkinsfile that is included in the pull request.

https://ci.jenkins.io/job/Plugins/job/docker-plugin/view/change-requests/job/PR-900/11/

**Comment by MarkEWaite** [comment]
@mat1e I would prefer that we first complete and merge:

* #923 

Once that is merged, then you will have the benefit of the bill of materials and already be based on Jenkins 2.346.x as minimum Jenkins version

**Comment by MarkEWaite** [comment]
@mat1e I've merged from the master branch into a copy of this branch, removed a few more items from the pom file, and made it visible at https://github.com/MarkEWaite/docker-plugin/commits/feature/from_netty_to_apachehttpclient

**Comment by mat1e** [comment]
> merged from the master branch into a copy of this branch, removed a few more items from the pom file, and made it visible at https://github.com/MarkEWaite/docker-plugin/commits/feature/from_netty_to_apachehttpclient

@MarkEWaite, It is ok for me if you want to replace this PR with your branch.

**Comment by shindouj** [comment]
Hello @MarkEWaite, @mat1e,

This PR looks so close to be merged - if neither of you can update this PR and/or create a new one based on Mark's branch, I am more than happy to click through it and create one for you all. Please let us know how do you want this deadlock to be resolved.

As a user that is blocked on ARM64 with this PR, I can't wait for this to be merged and deployed. :)

**Comment by shindouj** [comment]
> > Hello @MarkEWaite, @mat1e,
> > This PR looks so close to be merged - if neither of you can update this PR and/or create a new one based on Mark's branch, I am more than happy to click through it and create one for you all. Please let us know how do you want this deadlock to be resolved.
> > As a user that is blocked on ARM64 with this PR, I can't wait for this to be merged and deployed. :)
> 
> Are you willing to test the incremental build of the plugin?
> 
> * [Use Apache httpclient instead of netty #934](https://github.com/jenkinsci/docker-plugin/pull/934)

Sure, as long as you're not expecting a perfect QA process on my side I can check it on my Jenkins instance(s). :)

**Comment by MarkEWaite** [comment]
> Sure, as long as you're not expecting a perfect QA process on my side I can check it on my Jenkins instance(s). :)

Thanks for being willing to do that.  The pre-release build is https://ci.jenkins.io/job/Plugins/job/docker-plugin/job/PR-934/1/artifact/io/jenkins/docker/docker-plugin/1.2.11-rc1065.c3707157f5aa/docker-plugin-1.2.11-rc1065.c3707157f5aa.hpi 

Share your experiences with it as a comment here and that will be the catalyst to persuade me to merge it and release it.

**Comment by MarkEWaite** [comment]
Closing this pull request in favor of #934

**Comment by pjdarton** [comment]
Thank you (all) for persevering with this.
It was a long battle to get this working & I applaud your efforts. I'm sorry I had to abandon things (my own work pulled me away from here) and I'm glad to see that it worked out in the end.

**Comment by MarkEWaite** [comment]
> Thank you (all) for persevering with this. ... I'm glad to see that it worked out in the end.

Thanks so much for your efforts @pjdarton !  Your contributions to the plugin are greatly appreciated.  Any time you feel like contributing, we're thrilled to have whatever help you can offer.

**Pull request closed without merging**
